internal/fans/nvidia.go

```diff
@@ -0,0 +1,228 @@
+	// at least allows creating a fan curve
+}
+
+func (fan *NvidiaFan) GetRpmAvg() float64 {
+	return fan.RpmMovingAvg
+}
+
+func (fan *NvidiaFan) SetRpmAvg(rpm float64) {
+	fan.RpmMovingAvg = rpm
+}
+
+func (fan *NvidiaFan) GetPwm() (int, error) {
+	var device nvml.Device = nil // TODO!
+	fanIdx := fan.getNvFanIndex()
+	speed, ret := device.GetFanSpeed_v2(fanIdx)
+	if ret != nvml.SUCCESS {
+		speed = MinPwmValue // this is what HwMonFan does
+	}
+	// TODO: convert speed from percent (0..100) to PWM (0..255)?
+	return int(speed), nvError(ret)
+}
+
+func (fan *NvidiaFan) SetPwm(pwm int) (err error) {
+	ui.Debug("Setting Fan PWM of '%s' to %d ...", fan.GetId(), pwm)
+	var device nvml.Device = nil // TODO!
+	fanIdx := fan.getNvFanIndex()
+	// TODO: translate pwm (0..255) to percent (0..100)?
+	// or just clamp to 100 and let fan2go assume that it doesn't get faster after PWM > 100?
+	pwm = min(pwm, 100)
+	ret := device.SetFanSpeed_v2(fanIdx, pwm)
+	return nvError(ret)
+}
+
+func (fan *NvidiaFan) GetFanRpmCurveData() *map[int]float64 {
+	return fan.FanCurveData
+}
+
+// AttachFanCurveData attaches fan curve data from persistence to a fan
+// Note: When the given data is incomplete, all values up until the highest
+// value in the given dataset will be interpolated linearly
+// returns os.ErrInvalid if curveData is void of any data
+func (fan *NvidiaFan) AttachFanRpmCurveData(curveData *map[int]float64) (err error) {
+	if curveData == nil || len(*curveData) <= 0 {
+		ui.Error("Cant attach empty fan curve data to fan %s", fan.GetId())
+		return os.ErrInvalid
+	}
+
+	fan.FanCurveData = curveData
+
+	startPwm, maxPwm := ComputePwmBoundaries(fan)
+	fan.SetStartPwm(startPwm, false)
+	fan.SetMaxPwm(maxPwm, false)
+
+	// TODO: we don't have a way to determine this yet
+	fan.SetMinPwm(startPwm, false)
+
+	return err
+}
+
+func (fan *NvidiaFan) UpdateFanRpmCurveValue(pwm int, rpm float64) {
+	if fan.FanCurveData == nil {
+		fan.FanCurveData = &map[int]float64{}
+	}
+	(*fan.FanCurveData)[pwm] = rpm
+}
+
+func (fan *NvidiaFan) GetCurveId() string {
+	return fan.Config.Curve
+}
+
+func (fan *NvidiaFan) ShouldNeverStop() bool {
+	return fan.Config.NeverStop
+}
+
+func (fan *NvidiaFan) GetPwmEnabled() (int, error) {
+	var device nvml.Device = nil // TODO!
+	fanIdx := fan.getNvFanIndex()
+	policy, err := device.GetFanControlPolicy_v2(fanIdx)
+	if err != nvml.SUCCESS {
+		return 2, nvError(err)
+	}
+	pwm := 2 // "motherboard pwm control" as default assumption
+	if policy == nvml.FAN_POLICY_MANUAL {
+		if fan.RunAtMaxSpeed {
+			pwm = 0 // "max speed" - "no control" in hwmon backend
+		} else {
+			pwm = 1 // manual PWM control
+		}
+	}
+	return pwm, nil
+}
+
+func (fan *NvidiaFan) IsPwmAuto() (bool, error) {
+	value, err := fan.GetPwmEnabled()
+	if err != nil {
+		return true, err // assume auto control by default
+	}
+	return value == 2, nil
+}
+
+// SetPwmEnabled writes the given value to pwmX_enable
```

Comment:
the docs don't make sense here, and are also an indicator that the signature needs to be changed. This documentation is specific to the hwmon implementation.

Reply:
well, I try to emulate that here